servers/rendering/renderer_rd/shaders/tonemap.glsl

```diff
@@ -328,6 +329,55 @@ vec3 screen_space_dither(vec2 frag_coord) {
 	return (dither.rgb - 0.5) / 255.0;
 }
 
+// Adapted from https://github.com/DadSchoorse/vkBasalt/blob/b929505ba71dea21d6c32a5a59f2d241592b30c4/src/shader/cas.frag.glsl
+// (MIT license).
+vec3 do_cas(vec3 color, vec2 uv_interp, float sharpen_intensity) {
+	// Fetch a 3x3 neighborhood around the pixel 'e',
+	//  a b c
+	//  d(e)f
+	//  g h i
+	vec3 a = textureLodOffset(source_color, uv_interp, 0.0, ivec2(-1, -1)).rgb;
+	vec3 b = textureLodOffset(source_color, uv_interp, 0.0, ivec2(0, -1)).rgb;
+	vec3 c = textureLodOffset(source_color, uv_interp, 0.0, ivec2(1, -1)).rgb;
+	vec3 d = textureLodOffset(source_color, uv_interp, 0.0, ivec2(-1, 0)).rgb;
+	vec3 e = color.rgb;
+	vec3 f = textureLodOffset(source_color, uv_interp, 0.0, ivec2(1, 0)).rgb;
+	vec3 g = textureLodOffset(source_color, uv_interp, 0.0, ivec2(-1, 1)).rgb;
+	vec3 h = textureLodOffset(source_color, uv_interp, 0.0, ivec2(0, 1)).rgb;
+	vec3 i = textureLodOffset(source_color, uv_interp, 0.0, ivec2(1, 1)).rgb;
+
+	// Soft min and max.
+	//  a b c             b
+	//  d e f * 0.5  +  d e f * 0.5
+	//  g h i             h
+	// These are 2.0x bigger (factored out the extra multiply).
+	vec3 minRGB = min(min(min(d, e), min(f, b)), h);
+	vec3 minRGB2 = min(min(min(minRGB, a), min(g, c)), i);
+	minRGB += minRGB2;
+
+	vec3 maxRGB = max(max(max(d, e), max(f, b)), h);
+	vec3 maxRGB2 = max(max(max(maxRGB, a), max(g, c)), i);
+	maxRGB += maxRGB2;
+
+	// Smooth minimum distance to signal limit divided by smooth max.
+	vec3 rcpMaxRGB = vec3(1.0) / maxRGB;
+	vec3 ampRGB = clamp((min(minRGB, 2.0 - maxRGB) * rcpMaxRGB), 0.0, 1.0);
+
+	// Shaping amount of sharpening.
+	ampRGB = inversesqrt(ampRGB);
+	float peak = 8.0 - 3.0 * sharpen_intensity;
+	vec3 wRGB = -vec3(1) / (ampRGB * peak);
+	vec3 rcpWeightRGB = vec3(1.0) / (1.0 + 4.0 * wRGB);
+
+	//                          0 w 0
+	//  Filter shape:           w 1 w
+	//                          0 w 0
+	vec3 window = b + d + f + h;
+	vec3 outColor = clamp((window * wRGB + e) * rcpWeightRGB, 0.0, 1.0);
```

Comment:
This line clamps the output from 0.0 to 1.0.

Reply:
Replacing this with `max(vec3(0.0), …)` fixes the tonemapping issue, thanks :slightly_smiling_face: